Initial PageLoader Test ClassInitial test for the PageLoader class that covers null and empty checks

kennethmcfarland opened the pull request:
for factory methods and constructors.

kennethmcfarland committed:
Initial test for the PageLoader class that covers null and empty checks
for factory methods and constructors.

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+package webindex.data.fluo;
```
Need to add the apache file header.

kennethmcfarland replied:
Fixed

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+@SuppressWarnings("unused")
+public class PageLoaderTest {
+
+	@Test
```
Should this be `@Test(expected = NullPointerException.class)`?

keith-turner replied:
@kpm1985 do you know if this is still an issue?

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+	public void testDeleteFailsWithNullUrl() {
+		try {
+			PageLoader loader = PageLoader.deletePage(null);
+		} catch (MalformedURLException e) {
```
Can this catch be dropped?

kennethmcfarland replied:
I don't think so because of this method signature:

https://github.com/apache/fluo-examples/blob/47f1bb42969345b444f5c0e4c8c4a36ec93aab14/webindex/modules/data/src/main/java/webindex/data/fluo/PageLoader.java#L52

keith-turner replied:
If the exception is not expected you can make the test throw exception.  If it is actually thrown then the test will fail.

kennethmcfarland replied:
Not sure if this is correct but did a simple refactor where it catches Exception e and test if its an instance of a NPE and if yes throws NPE.

keith-turner replied:
I was thinking of the following.

```java
@Test(expected = NullPointerException.class)
public void testDeletePageThrowsNPEifNullURL() throws MalformedURLException {	
  PageLoader.deletePage((URL)null);
}
```

kennethmcfarland replied:
Ok, I gotcha now. I fixed it to throws instead of catch, and did the same for the other methods. I added throws to the last test for clarity even though it catches all exceptions, as a hint to the reader.

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+	public void testDeletePage() {
+		URL url = new URL("","","",0,false,false);
+		try {
+			PageLoader loader = PageLoader.deletePage(url);
```
What is the intent of this test?

kennethmcfarland replied:
Try to see if deletePage() works with an empty URL. Will refactor with a better method name.

kennethmcfarland replied:
The method has been given a vastly better descriptive name that does not require guesswork.

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+		}
+	}
+	
+	@Test
```
Should this be `@Test(expected = NullPointerException.class)`?  Also if that change is made seem like all of the catch blocks could be dropped.

keith-turner replied:
@kpm1985 do you know if this is still an issue?

ctubbsii commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,63 @@
+import webindex.core.models.Page;
+import webindex.core.models.URL;
+
+@SuppressWarnings("unused")
```
There's no reason to suppress this warning, since this is easily avoidable. For example, replace:
```java
Page p = null;
PageLoader loader = PageLoader.updatePage(p);
```

with:

```java
PageLoader loader = PageLoader.updatePage((Page) null);
```

kennethmcfarland replied:
Thank you, fixed.

kennethmcfarland committed:
Refactoring of PageLoaderTest

Added Apache License header
Renamed methods with better descriptives
Removed unused local variable warnings

kpm1985 commented:
Just to document a few things and why this class/test even exists, I was reading through the code and I noticed it had an empty check but not a null check. Then instead of just saying that, I thought it would be good to write a test that exposes that problem. 

I personally wanted to get more used to (experience) writing JUnit tests so I did a pretty quick hack job to be a work in progress.  I'm also working being a lot more active than I have been and it seemed like a fine idea to write some tests and increase the test coverage. 

As for the PageLoader class, I noticed load(Tx, Ctx) doesnt check for null args.  The deletePage(URL url) requires url not be null but doesn't care if its a url representing "", ie empty or blank. The updatePage(Page page) method requires the url is not "" or blank/empty but doesn't care if it gets a null page (leading to NPE).

So I was like sweet, maybe this is a decent class for me to write a test for. I did one refactoring based on the first feedback, thank you for the feedback! Casting null is something I'm not even sure I've seen before, that was really neat.

There is still some to be done with the test as the load(tx,ctx) method uses deletePage(URL url) and I am currently passing an empty page to it which fails the deletePage(URL) method but can just produce a real non malformed url like maybe new URL("http://fluo.apache.org")

kennethmcfarland committed:
Update Exception Handling

Improve handling of unexpected exceptions and cast them to the
appropriate type

kpm1985 commented:
I created a separate issue for the refactoring of PageLoader so that it can pass these tests.

kennethmcfarland committed:
Update Exception Handling

Change method signature to eliminate catch blocks

keith-turner commented on webindex/modules/data/src/test/java/webindex/data/fluo/PageLoaderTest.java:
```diff
@@ -0,0 +1,62 @@
+			loader.load(null, null);
+		} catch (Exception e) {
+			if(e instanceof NullPointerException)
+				throw new NullPointerException(e.getMessage());
```
I don't understand why this is catching and then rethrowing a NPE.  Why even catch it?

kennethmcfarland replied:
Hello Keith, I did that because loader class throws Exception, but I was trying to determine when it's because of a NullPointer to load(tx,ctx) rather than just the exception load already throws, I fixed it and removed it.

kennethmcfarland committed:
Update exception handling method signature

keith-turner commented:
@kpm1985 the changes are looking good now.  So does this test not pass as it is?

kpm1985 commented:
That is correct, I just run it in Eclipse and get a couple failing tests.

kpm1985 commented:
@keith-turner The two tests that are failing are testUpdatePageWithNullPage and testLoadWithNullTxCtx. 
The first failing test comes from here https://github.com/apache/fluo-examples/blob/47f1bb42969345b444f5c0e4c8c4a36ec93aab14/webindex/modules/data/src/main/java/webindex/data/fluo/PageLoader.java#L44-L50

Where I think a simple Objects.requireNonNull(page) could do the trick.

The second comes from here: https://github.com/apache/fluo-examples/blob/47f1bb42969345b444f5c0e4c8c4a36ec93aab14/webindex/modules/data/src/main/java/webindex/data/fluo/PageLoader.java#L60-L80

And Objects.requireNonNull() could again do the trick.

kpm1985 commented:
@keith-turner @ctubbsii Anybody have some feedback? Does this look ok now?
